src/kernel/scheduler.zig

```diff
@@ -0,0 +1,583 @@
+    ///
+    /// Arguments:
+    ///     IN/OUT self: *Self - The pointer to self.
+    ///
+    pub fn destroy(self: *Self) void {
+        freePid(self.pid);
+        alloc.free(self.stack);
+        alloc.destroy(self);
+    }
+
+    ///
+    /// Save a new stack pointer for this task.
+    ///
+    /// Arguments:
+    ///     IN/OUT self: *Self - The pointer to self.
+    ///     IN esp: u32        - The new stack pointer to be saved.
+    ///
+    pub fn saveEsp(self: *Self, esp: u32) void {
+        self.esp = esp;
+    }
+};
+
+///
+/// Allocate a process identifier. If out of PIDs, then will panic. Is this occurs, will need to
+/// increase the bitmap.
+///
+/// Return: u32
+///     A new PID.
+///
+fn allocatePid() u32 {
+    return all_pids.setFirstFree() orelse panic(@errorReturnTrace(), "Out of PIDs", .{});
+}
+
+///
+/// Free an allocated PID. One must be allocated to be freed. If one wasn't allocated will panic.
+///
+/// Arguments:
+///     IN pid: u32 - The PID to free.
+///
+fn freePid(pid: u32) void {
+    all_pids.clearEntry(pid) catch panic(@errorReturnTrace(), "PID {} not allocated", .{pid});
+}
+
+///
+/// Round robin. This will pick the next task to be run from the queue. It will add the current
+/// task to the end of the queue and pop the next task from the front as set this as the current
+/// task. Interrupts are assumed disabled.
+///
+fn pickNextTask() void {
+    // Get the next task
+    if (tasks.pop()) |new_task_node| {
+        var next_task = new_task_node.data;
+        tasks.destroyNode(new_task_node, alloc);
+
+        // Now that we have the new task and deleted its node, there "will" be space for another
+        // so the allocation shouldn't fail as interrupts are disabled.
+        // So if there is a failure, then wow :P This will be the case for multiprocessor, but
+        // for a later time
+
+        // Put the current task back in the queue, at the back
+        var cur_elem = tasks.createNode(current_task, alloc) catch |e| {
+            panic(@errorReturnTrace(), "Error in pickNextTask, can't save previous task: {}\n", .{e});
+        };
+        tasks.prepend(cur_elem);
+
+        // Set the next task
+        current_task = next_task;
+    } else {
+        // Not more tasks, this isn't possible as the idle task is noreturn
+        panic(@errorReturnTrace(), "No more tasks?\n", .{});
+    }
+}
+
+///
+/// The idle task that just halts the CPU but the CPU can still handle interrupts.
+///
+fn idle() noreturn {
+    arch.spinWait();
+}
+
+///
+/// Create a new task and add it to the scheduling queue. No locking.
+///
+/// Arguments:
+///     IN comptime entry_point: var - The entry point into the task. This must be a function.
+///
+/// Error: std.mem.Allocator.Error
+///     OutOfMemory - If there isn't enough memory for the a task/stack. Any memory allocated will
+///                   be freed on return.
+///
+pub fn createNewTask(comptime entry_point: var) Allocator.Error!void {
```

Comment:
Perhaps this function should be made to just add to the task queue since task initialisation is done with `Task.init()` anyway. If so then a suitable new name for it could be `schedule`.

Reply:
Not really sure what you mean